File: uni_modules/ui-component/utssdk/app-ios/swift/observer/LiveAudienceStoreObserver.swift
import Combine
import DCloudUTSFoundation
import RTCRoomEngine_Plus

public class LiveAudienceStoreObserver {
    private var cancellables = Set<AnyCancellable>()
    public static let shared = LiveAudienceStoreObserver()
    
    public func liveAudienceStoreChanged(_ liveId: String, _ callback: @escaping (_ name: String, _ data: String) -> Void) {
        cancellables.removeAll()

        V2RTCRoomEngine.shared.getLiveAudienceStore(liveId: liveId)
            .state.subscribe(StateSelector(keyPath: \V2LiveAudienceState.audienceList))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] audienceList in
                guard let self = self else { return }
                var list: [[String: Any]] = []
                for info in audienceList {
                    list.append(convertAudienceInfoToDic(info: info))
                }
                if let jsonList = JsonUtil.toJson(list) {
                    callback("audienceList", jsonList)
                }
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getLiveAudienceStore(liveId: liveId)
            .state.subscribe(StateSelector(keyPath: \V2LiveAudienceState.audienceListCursor))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] cursor in
                guard let self = self else { return }
                if let cursor = cursor {
                    callback("audienceListCursor", String(cursor))
                } else {
                    callback("audienceListCursor", "")
                }
            })
            .store(in: &cancellables)
    }

    private func convertAudienceInfoToDic(info: AudienceInfo) -> [String: Any] {
        var dict: [String: Any] = [
            "userId": info.userId,
            "userName": info.userName,
            "avatarUrl": info.avatarUrl,
            "customInfo": info.customInfo,
            "userRole": TypeConvert.convertUserRole(info.userRole),
            "isMessageDisabled": info.isMessageDisabled,
            "joinedTimestamp": info.joinedTimestamp,
        ]
        return dict
    }   
}

File: uni_modules/ui-component/utssdk/app-android/kotlin/observer/LiveAudienceStoreObserver.kt
package uts.sdk.modules.uiComponent.observer

import com.google.gson.Gson
import com.tencent.cloud.uikit.state.AudienceInfo
import com.tencent.cloud.uikit.state.V2RTCRoomEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import uts.sdk.modules.uiComponent.kotlin.observer.TypeConvert

object LiveAudienceStoreObserver {
    private val gson = Gson()
    private var bindDataJob: Job? = null

    fun liveAudienceStoreChanged(liveId: String, callback: (String, String) -> Unit) {
        bindDataJob?.cancel()
        bindDataJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                V2RTCRoomEngine.getInstance()
                    .getLiveAudienceStore(liveId).liveAudienceState.audienceList.collect { audienceList ->
                        val list = audienceList?.map { convertAudienceInfoToMap(it) }
                        callback("audienceList", gson.toJson(list))
                    }
            }

            launch {
                V2RTCRoomEngine.getInstance().getLiveAudienceStore(liveId).liveAudienceState.audienceListCursor
                    .collect { cursor ->
                        callback("audienceListCursor", gson.toJson(cursor))
                    }
            }

        }
    }

    private fun convertAudienceInfoToMap(info: AudienceInfo): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        map["userId"] = info.userId
        map["userName"] = info.userName
        map["avatarUrl"] = info.avatarUrl
        map["customInfo"] = info.customInfo
        map["userRole"] = TypeConvert.convertUserRole(info.userRole)
        map["isMessageDisabled"] = info.isMessageDisabled
        map["joinedTimestamp"] = info.joinedTimestamp
        return map
    }
} 